runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,84 @@
+local G = require('_snippet_grammar')
+
+--- Returns the current cursor position (0-based).
+--- @return { line: integer, col: integer }
+local function cursor_position()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  return { line = cursor[1] - 1, col = cursor[2] }
+end
+
+--- Resolves variables (like `$name` or `${name:default}`) as follows:
+--- - When a variable is unknown (i.e.: its name is not recognized in any of the cases below), return `nil`.
+--- - When a variable isn't set, return its default (if any) or an empty string.
+---
+--- Note that in some cases, the default is ignored since it's not clear how to distinguish an empty
+--- value from an unset value (e.g.: `TM_CURRENT_LINE`).
+---
+--- @param var string
+--- @param default? string
+--- @return integer|string|nil
+local function resolve_variable(var, default)
+  if var == 'TM_SELECTED_TEXT' then
+    return default or ''
+  elseif var == 'TM_CURRENT_LINE' then
+    return vim.api.nvim_get_current_line()
+  elseif var == 'TM_CURRENT_WORD' then
+    return vim.fn.expand('<cword>') --[[@as string]]
+      or default
+  elseif var == 'TM_LINE_INDEX' then
+    return cursor_position().line
+  elseif var == 'TM_LINE_NUMBER' then
+    return cursor_position().line + 1
+  end
+
+  local bufname = vim.api.nvim_buf_get_name(0)
+  if bufname == '' then
+    return default
+  end
+  if var == 'TM_FILENAME' then
+    return vim.fn.fnamemodify(bufname, ':p:t')
+  elseif var == 'TM_FILENAME_BASE' then
+    local filename_base = vim.fn.fnamemodify(bufname, ':p:t'):gsub('%.[^%.]*$', '')
+    return filename_base
+  elseif var == 'TM_DIRECTORY' then
+    return vim.fn.fnamemodify(bufname, ':p:h:t')
+  elseif var == 'TM_FILEPATH' then
+    return bufname
+  end
+
+  -- Unknown variable.
+  return nil
+end
+
+--- @alias vim.snippet.Direction
+--- | -1 # Previous
+--- | 1 # Next
+
+local M = {}
+
+--- @param direction vim.snippet.Direction
+function M.jumpable(direction) end
+
+--- @param direction vim.snippet.Direction
+function M.jump(direction) end
+
+--- Expands the given snippet text.
+--- @param snippet_text string
+function M.expand(snippet_text)
+  local snippet = G.parse(snippet_text)
+
+  for _, child in ipairs(snippet.data.children) do
+    local type, data = child.type, child.data
+    if type == G.NodeType.Variable then
+      --- @cast data vim.snippet.VariableData
+      -- Try to get the variable's value.
+      local value = resolve_variable(data.name, tostring(data.default))
```

Comment:
Note that the snippet grammar says that the variable's default can be any node, but here's I'm just assuming that their string representation should be inserted and not doing any extra parsing.